src/cargo/util/machine_message.rs

```diff
@@ -0,0 +1,30 @@
+use rustc_serialize::json;
+use core::{PackageId, Target};
+
+#[derive(RustcEncodable)]
+pub struct FromCompiler<'a> {
+    reason: &'static str,
+    package_id: &'a PackageId,
+    target: &'a Target,
+    message: json::Json,
+}
+
+impl<'a> FromCompiler<'a> {
+    pub fn new(package_id: &'a PackageId,
+               target: &'a Target,
+               message: json::Json)
+               -> FromCompiler<'a> {
+        FromCompiler {
+            reason: "compiler-message",
+            package_id: package_id,
+            target: target,
+            message: message,
+        }
+    }
+
+    pub fn emit(self) {
+        let json = json::encode(&self).unwrap();
+        println!("{}", json);
```

Comment:
@alexcrichton is this the right way to print JSON to stdout? There is no error reporting, and it bypasses config.shell altogether, but I think both of these are OK: we don't use colors, so we don't need shell, `println` does necessary synchronization, and it is probably OK to panic if we fail to write to stdout.